src/utils/request.js

```diff
@@ -112,15 +114,21 @@ class Request {
    * @param {string} opts.endpoint Upload endpoint
    * @param {ReadableStream} opts.dataStream Readable data stream
    * @param {number} opts.dataLength Data length in bytes
+   * @param {string} [opts.query] Querystring key-value parameters
    * @returns {Promise<Object>} Promise representing the upload result
    */
-  async _upload({ endpoint, dataStream, dataLength }) {
+  async _upload({ endpoint, dataStream, dataLength, query = {} }) {
+    const uploadUrl = new URL(`${this._baseUrl}${endpoint}`);
+    Object.keys(query).forEach(
+      (key) => query[key] && uploadUrl.searchParams.append(key, query[key])
+    );
```

Comment:
Can we just use `URLSearchParams`?

```suggestion
    uploadUrl.searchParams = new URLSearchParams(query);
```

Reply:
This won't work as expected.

Reply:
`{ exp: undefined }` will produce `?exp=undefined` ;)